recipes/logr/0.1.0/conanfile.py

```diff
@@ -0,0 +1,92 @@
+from conans import ConanFile, CMake, tools
+from conans.errors import ConanInvalidConfiguration
+import os
+
+class LogrConan(ConanFile):
+    name = "logr"
+    version = "0.1.0"
+    license = "BSD 3-Clause License"
+    homepage = "https://github.com/ngrodzitski/logr"
+    url = "https://github.com/conan-io/conan-center-index"
+    description = "Logger frontend substitution for spdlog, glog, etc for server/desktop applications"
+    topics = ("logger", "development", "util", "utils")
+    generators = "cmake"
+    settings = "os", "compiler", "build_type", "arch"
+    build_policy = "missing"
+    exports_sources = ["CMakeLists.txt"]
+
+    options = { 'spdlog_backend' : [True, False],
+                'glog_backend' : [True, False],
+                'log4cplus_backend' : [True, False] }
+
+    default_options = { 'spdlog_backend': True,
+                        'glog_backend': True,
+                        'log4cplus_backend': True }
+
+    _cmake = None
+
+    @property
+    def _source_subfolder(self):
+        return "source_subfolder"
+
+    @property
+    def _build_subfolder(self):
+        return "build_subfolder"
+
+    def requirements(self):
```

Comment:
I'm surprised you don't add a requirement on either spdlog, glog or log4plus.
If this is a frontend substitution, then the backend should be available. Isn't it?

Reply:
I consider the following reasoning: 

* Deciding what logger to use in your application (among the three supported).
* Once logger library is decided, the application adds it to the list of dependencies (so logger is already in deps list).
* Then it is decided to use another "frontend" to it.

Deciding the logger happens in the first place (so deps machinery is applied for logger lib), and the logr is like an extension to a selected library.

And in case of using in conanfile it would be less bothering to just specify logr without forcing the user to dig into options.

Reply:
The problem is that  `logr` won't work when having the following conanfile.txt:
```
[requires]
spdlog/0.1.0
[options]
logr:with_spdlog_backend=True
logr:with_glog_backend=False
logr:with_log4cplus_backend=False
```
because spdlog is is not available.
I would do this in `requirements`:
```python
if self.options.with_spdlog_backend:
    self.requires("spdlog/x.y")
elif self.options.with_glog_backend
    self.requires("log4cplus/x.y")
elif self.options.with_log4cplus_backend:
    self.requires("log4cplus/x.y")
```

Reply:
I've test the following conanfile.txt
```
[requires]
logr/0.1.0
[options]
logr:spdlog_backend=True
logr:glog_backend=False
logr:log4cplus_backend=False

[generators]
cmake
```
And a project based on test_project works just fine.

The libs is a collection of headers and it has some options to add a certain headers or not.

Would it be ok to just remove all options and make all headers to be unconditionally present?

Reply:
How usable is this library without any backend?

Reply:
If someone provides his own backend, then it is also a way to go.
Are you suggesting a single option e.g. "backend" with values "spdlog", "glog", "log4cplus", None?

Reply:
That could be a way, but I think you should also add these library as a requirement of logr. Or shouldn't you?

Reply:
```
    options = { "backend": ["spdlog", "glog", "log4cplus", "all", None }
    default_options = { "backend": ["spdlog"]}
```
all is the option for local dev builds. As I want all of them to be available in single build (for examples and benchmarks) and I want to keep local conanfile at most similar to projects own conanfile.

Does Conan introduces dependencie propogation when I specify `self.requires("spdlog/x.y")`? Or I need to modify projects cmake files and add something like `target_link_libraries(logr INTERFACE spdlog::spdlog)`?

Reply:
Personally, I have given up on keeping project's conanfiles equal to cci's conanfiles.
Mostly because I don't want the project to need a cmake wrapper script, just for conan.

When you add `self.requires("spdlog/x.y")` to logr,
then linking to the `logr::logr` cmake target will also link to `spdlog::spdlog`.

I changed your example a bit by removing the list from `default_options`.
It can only have one.
```python
    options = { "backend": ["spdlog", "glog", "log4cplus", "all", None }
    default_options = { "backend": "spdlog"}
```
As @prince-chrismc already mentioned, I don't think an `all` option does not make muck sense for cci. For your personal development branch, it might.

Reply:
> Personally, I have given up on keeping project's conanfiles equal to cci's conanfiles.

Yep! CCI is trying to provide the most convince to as many consumers as possible. Which may not meet the needs of the project!

That being said, as a very happy RESTinio consumer, it seems unlikely that a project will have multiple logging methods.... but then they wouldn't need this library either! 

✅ There is no right or wrong answer, so ultimately the choice is yours. 

I have to agree with @madebr suggestion for a single option that adds the correct requires... for myself this adds the freedom to change in the future with little impact.

Again the opinion of a potential consumer =)

Reply:
Refactored to a single option "backend", which implies dependency to a specific logger library.